pkg/resources/reconciler.go

```diff
@@ -228,47 +228,82 @@ func (r *Reconciler) ReconcileSubscription(ctx context.Context, target marketpla
+	if ip.Status.Phase == operatorsv1alpha1.InstallPlanPhaseFailed {
+		var csv *operatorsv1alpha1.ClusterServiceVersion
+		if sub.Status.InstalledCSV != "" {
+			csv = &operatorsv1alpha1.ClusterServiceVersion{
+				ObjectMeta: metav1.ObjectMeta{Namespace: target.Namespace, Name: sub.Status.InstalledCSV},
 			}
-			log.Warningf("Deleting subscription for re-install due to failed install plan", l.Fields{"ns": target.Namespace, "install plan": target.Pkg})
-			if err := client.Delete(ctx, sub); err != nil && !k8serr.IsNotFound(err) {
-				return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("failed to delete subscription for re-intstall: %s", err)
+		}
+		return retryInstallation(ctx, client, log, target, csv, sub)
+	}
+
+	//if it's approved but not complete, then it's in progress
+	if ip.Status.Phase != operatorsv1alpha1.InstallPlanPhaseComplete && ip.Spec.Approved {
+		log.Infof("Install plan is not complete yet ", l.Fields{"install plan": target.Pkg})
+		return integreatlyv1alpha1.PhaseInProgress, nil
+		//if it's not approved by now, then it will not be approved by this version of the integreatly-operator
+	} else if !ip.Spec.Approved {
+		log.Infof("Upgrade installplan above the maximum allowed version", l.Fields{"install plan": target.Pkg})
+	}
+
+	for _, csvName := range ip.Spec.ClusterServiceVersionNames {
+		ipCSV := &operatorsv1alpha1.ClusterServiceVersion{}
+		if err := client.Get(ctx, k8sclient.ObjectKey{
+			Name:      csvName,
+			Namespace: target.Namespace,
+		}, ipCSV); err != nil {
+			if k8serr.IsNotFound(err) {
+				log.Infof("Waiting for CSV to be created in cluster after InstallPlan is complete: %s", l.Fields{
+					"install plan": target.Pkg,
+				})
+				return integreatlyv1alpha1.PhaseInProgress, nil
 			}
-			return integreatlyv1alpha1.PhaseAwaitingOperator, nil
+			return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("error retrieving CSV: %v", err)
 		}
 
-		//if it's approved but not complete, then it's in progress
-		if ip.Status.Phase != operatorsv1alpha1.InstallPlanPhaseComplete && ip.Spec.Approved {
-			log.Infof("Install plan is not complete yet ", l.Fields{"install plan": target.Pkg})
-			return integreatlyv1alpha1.PhaseInProgress, nil
-			//if it's not approved by now, then it will not be approved by this version of the integreatly-operator
-		} else if !ip.Spec.Approved {
-			log.Infof("Upgrade installplan above the maximum allowed version", l.Fields{"install plan": target.Pkg})
+		if err := validateCSV(ipCSV); err != nil {
+			log.Warningf("CSV failed validation. Retrying operator installation", l.Fields{"error": err, "install plan": target.Pkg})
+			return retryInstallation(ctx, client, log, target, ipCSV, sub)
 		}
 	}
 
 	return integreatlyv1alpha1.PhaseCompleted, nil
 }
 
+func validateCSV(csv *operatorsv1alpha1.ClusterServiceVersion) error {
+	if csv.Spec.InstallStrategy.StrategyName == operatorsv1alpha1.InstallStrategyNameDeployment && len(csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs) == 0 {
```

Comment:
Hey @sergioifg94 , so just to confirm, is the issue with the current install plans that they might have a deployment specified but no actual deployment specs inside them?

Reply:
Hey Joseph, so there's actually 2 issues addressed in this PR: 1. The duplicated install plans and 2. The possibility of an install plan having no deployment. I believe the second one is indirectly caused the first one, but for safety I've addressed both (that's why they're in 2 separate commits). 
The second issue seems to be caused by the install plan not including a deployment at all in the CSV. A CSV may include many deployments, that's why I'm checking that the length of the array is not zero.

Here's an example of a CSV with 2 deployments: https://github.com/integr8ly/integreatly-operator/blob/master/manifests/integreatly-marin3r/0.7.0/marin3r.v0.7.0.clusterserviceversion.yaml#L481-L559